### DISCUSSION TITLE
Trying to Understand Code

### DISCUSSION THREAD

**Discussion by sandstone42** [GrandOrgue/grandorgue#938] [General]
I know this is something others have been trying to do. Since there is no general outline indicating how classes are instantiated the only way I've found is to take the time to examine one cpp at a time (for example, GOOrgan.cpp) and search the code for where the GOOrgan constructor is instantiated and then where the various methods for GOOrgan are used. It takes a lot of time and I, for one, lose interest very quickly. Anyone else have a better method?

**Comment by jason-barba** [comment]
https://en.wikipedia.org/wiki/Unified_Modeling_Language ?

**Comment by larspalo** [comment]
As much as I want to see more people involved in GrandOrgue coding, I also realize that it's a huge process to learn programming both the basic c++ and the wxwidgets framework upon that and then all the external libraries that GO uses. Thus I've come to the conclusion that only the "crazy" (enough) people that don't find this process daunting should even bother... It's much better if you actually "enjoy" reading and writing code! The GrandOrgue project don't only need programming "nerds" but also users and sampleset producers on all levels to really prosper. So, in short, if you're not feeling coding is in your bag of tricks - don't feel sorry for that! Enjoy playing GO and see if there are other aspects of the project that you feel you can help with, that's absolutely valuable enough!

I can though, share a little about my own journey into programming. In "ancient times", when I was about 10 years old I got my first computer that I learned programming on - first Basic, later DOS and Turbo Pascal. Then I got really into music and from about 15 years old I already knew that I was going to be an Organist and work in the church. So, at that time I basically only focused on music and played/practised all my free time and forgot about computers...

Then around 2005 I was so fed up with Windows blue screens that I decided to search for an alternative OS - enter Linux! Around that time I only wrote code in HTML/XHTML, CSS, MySQL and PHP, but within a few years I discovered GENPO, Aeolus and jOrgan and was very fortunate to come into contact with Graham Goode that introduced me to the MyOrgan project that I ran in wine... Shortly after that the GrandOrgue project was born and I got involved relatively early on, first as regular user, then tester and sampleset developer - first as a co worker with Graham on Jeuxd'orgues2.

In that process I realized that I should learn more about programming, partly to help with GO but also to get a decent software for sample production on Linux. LoopAuditioneer dates back to 2011 officially, but I started developing it a bit earlier. I decided to use the same kind of tools that GO uses, c++, wxwidgets and RtAudio. But also I learned C because of the dependency on libsndfile. Basically, I had a problem that I needed to solve and from that need all the learning happened.

I have been extremely fortunate to have had lots of help from Graham Goode, @nickappleton (also with his noise reduce program that all my free samplesets have been processed through) and Martin Koegler all the way. I've read quite a few books on java, c, c++ and wxwidgets and I have written many small programs to learn the basics, and LoopAuditioneer was my first more ambitious project that I worked on at the same time as I tested the new GO features and slowly learned to follow along in the code...

Even after all the years I've now spent with GO code I still don't know all of it by heart, but I do know enough that I can track down the flow of things and know where to look when something needs addressing. There are always more things to learn but I really do enjoy the process of learning and writing code is something that makes me happy... I'm still first and foremost a musician by profession and should most likely still only be classed as a hobby programmer!

So, my advice is really - don't start with GO code only to begin with! Start with smaller projects and learn the basics of c++ and wxwidgets. Then work your way into GO code if you're so inclined. GO code flow starts in src/grandorgue/GOApp.h and src/grandorgue/GOApp.cpp.

**Reply by paolo2504** [reply]
Lars, thanks for sharing this with all of us!
